### PR TITLE
Force babel helpers

### PR DESCRIPTION
**Pull request by Andarist**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,31 +1,40 @@
...
-import { addBabelPlugin, escapeRegExpCharacters, warnOnce } from './utils.js';
-import { RUNTIME, EXTERNAL, HELPERS } from './constants.js';
+import preflightCheck from './preflightCheck.js';
+import bundledHelpersPlugin from './bundledHelpersPlugin.js';
+import { addBabelPlugin, escapeRegExpCharacters } from './utils.js';
+import { RUNTIME, EXTERNAL, BUNDLED, INLINE, HELPERS } from './constants.js';
 
 const unpackOptions = ({
 	extensions = babel.DEFAULT_EXTENSIONS,
+	skipBabelHelpersCheck = false,
```
Comment (Andarist) [inline src/index.js]: this is not quite true - I also ignore module check with this flag, need to rename it

**Comment by christophehurpeau** [comment]
Hello @Andarist, do you need help to move forward this pr ?

**Comment by Andarist** [comment]
@christophehurpeau kinda, would be good to at least get a second opinion about this change

**Comment by christophehurpeau** [comment]
I tested this version on several projets with
```
skipBabelHelpersCheck: true,
babelHelpers: 'runtime',
```

it works perfectly :)

File: src/index.js [hunk 2]
```diff
@@ -71,15 +73,15 @@ function createBabelPluginFactory(customCallback = returnObject) {
 					return;
 				}
 
-				return babel.buildExternalHelpers(externalHelpersWhitelist, 'module');
+				return babel.buildExternalHelpers(null, 'module');
```
Comment (skeggse) [inline src/index.js]: What's the rationale for dropping `externalHelpersWhitelist` support? Is it never used in the wild, or no longer valid? I'm all for simplicity here, so :+1: from me :)

File: src/preflightCheck.js [hunk 3]
```diff
@@ -25,42 +25,46 @@ function fallbackClassTransform() {
...
 }
 
-export default function createPreflightCheck() {
-	let preflightCheckResults = {};
+const PREFLIGHT_INPUT = 'class Foo extends Bar {};\nexport default Foo;';
 
-	return (ctx, options) => {
-		const key = options.filename;
+const mismatchError = (actual, expected, filename) =>
+	`You have declared using "${expected}" babelHelpers, but transforming ${filename} resulted in "${actual}. Please check your configuration."`;
```
Comment (skeggse) [inline src/preflightCheck.js]:
```suggestion
	 `You have declared using "${expected}" babelHelpers, but transforming ${filename} resulted in "${actual}". Please check your configuration.`;
```

File: test/test.js [hunk 4]
```diff
@@ -56,7 +56,7 @@ describe('rollup-plugin-babel', function() {
 				Object.assign(
 					{
 						input,
-						plugins: [babelPlugin(babelOptions)],
+						plugins: [babelPlugin(Object.assign({ babelHelpers: 'bundled' }, babelOptions))],
```
Comment (skeggse) [inline test/test.js]:
```suggestion
						plugins: [babelPlugin({ babelHelpers: 'bundled', ...babelOptions })],
```

File: src/index.js [hunk 5]
```diff
@@ -1,31 +1,40 @@
...
-	...rest,
-	caller: {
-		name: 'rollup-plugin-babel',
-		supportsStaticESM: true,
-		supportsDynamicImport: true,
-		...rest.caller,
-	},
-});
+} = {}) => {
+	if (!rest.babelHelpers) {
```
Comment (skeggse) [inline src/index.js]: I'm tempted to say we should preserve some of the auto-detection functionality. It's super useful in two cases: 1) when you're getting started with rollup or a new project, and 2) for small projects where you want it to Just Work™. Larger projects definitely benefit from explicitly setting this.

One proposal: measure the number of times we run the preflight check for auto-detection and if it exceeds a certain number (or amount of time elapsed) then log a warning to the console asking the developer to specify the `babelHelpers` key (which they can choose to just be `auto` or something).

File: test/test.js [hunk 6]
```diff
@@ -188,21 +198,10 @@ describe('rollup-plugin-babel', function() {
...
-		console.warn = msg => {
-			messages.push(msg);
-		};
-		return bundle('samples/external-helpers-deprecated/main.js').then(() => {
-			console.warn = originalWarn;
-
-			assert.deepEqual(messages, [
-				'Using "external-helpers" plugin with rollup-plugin-babel is deprecated, as it now automatically deduplicates your Babel helpers.',
-			]);
+	it('allows using in combination with @babel/plugin-external-helpers', () => {
```
Comment (skeggse) [inline test/test.js]: Oh neat! Can you walk me through how these changes result in us supporting `@babel/plugin-external-helpers` again? I'm far less familiar with this use-case.

File: src/index.js [hunk 7]
```diff
@@ -1,31 +1,40 @@
...
+		throw new Error(
+			'You have to specify how do you want to bundle/import "Babel helpers" (runtime functions inserted by Babel which are used by some transformations).\n\n' +
+				'Please pass `babelHelpers` option to the rollup-plugin-babel with one of the following values:\n' +
+				`  - "${RUNTIME}" - you should use it especially when building libraries with rollup. It has to be used in combination with \`@babel/plugin-transform-runtime\` and you should also specify \`@babel/runtime\` as dependency of your package (don't forget to tell rollup to treat it is your external dependency when bundling for cjs & esm formats).\n` +
```
Comment (skeggse) [inline src/index.js]: Would it be possible (or even advisable) to auto-add `@babel/plugin-transform-runtime` if not included?

File: src/preflightCheck.js [hunk 8]
```diff
@@ -25,42 +25,46 @@ function fallbackClassTransform() {
...
-			}
-
-			if (
-				!~check.indexOf('export default') &&
-				!~check.indexOf('export default Foo') &&
-				!~check.indexOf('export { Foo as default }')
-			) {
-				ctx.error(MODULE_ERROR);
-			}
+	if (check.match(/\/helpers\/(esm\/)?inherits/)) {
```
Comment (skeggse) [inline src/preflightCheck.js]:
```suggestion
	if (/\/helpers\/(esm\/)?inherits/.test(check)) {
```

File: src/index.js (hunk 2 above)
Comment (Andarist) [reply src/index.js]: I'm aiming for simplicity here + rollup should be able to tree shake the generated module, so dont think that there is a strong need for this whitelist.

File: test/test.js (hunk 4 above)
Comment (Andarist) [reply test/test.js]: This plugin is bundled by buble instead of babel (couldnt figure out building with itself) and I had some issues with transforming spread operator with buble, but gonna give it a try.

File: test/test.js (hunk 6 above)
Comment (Andarist) [reply test/test.js]: `@babel/plugin-external-helpers` assumes that helpers are available on `babelHelpers` object and that it's reachable in the final scope (so most commonly - that it is a global), it's probably relict of the past and not used by many, but at least we can bring "support" back for it

File: src/index.js (hunk 7 above)
Comment (Andarist) [reply src/index.js]: Installing dependencies is rather beyond the scope of this plugin, but maybe we could at least validate if it's installed properly.

File: src/preflightCheck.js (hunk 8 above)
Comment (Andarist) [reply src/preflightCheck.js]: 👌

File: src/index.js (hunk 7 above)
Comment (skeggse) [reply src/index.js]: Ah yeah that's reasonable! I was more suggesting we could insert the plugin into the list, but just checking for validity is fine :+1:

File: test/test.js (hunk 4 above)
Comment (skeggse) [reply test/test.js]: :joy: self-hosting is always fun

File: src/index.js (hunk 2 above)
Comment (skeggse) [reply src/index.js]: makes sense to me, thanks!

File: src/index.js (hunk 5 above)
Comment (Andarist) [reply src/index.js]: My main motivation behind forcing this config on people was that people are often not aware that they "should" use runtime helpers.

I.e. since webpack@4 it requires "mode" config, however it defaults it to "production". Maybe we could do similarly here - leave current logic (no config, bundled helpers) as default BUT give a short warning (with a link to a full explanation) about configuring this explicitly. WDYT?

Comment (skeggse) [reply src/index.js]: Warning: wall of text incoming. I wanted to share most of my thought process, but see TLDR at bottom if you don't care :joy:

>My main motivation behind forcing this config on people was that people are often not aware that they "should" use runtime helpers.

Absolutely agreed.

I think in many ways forcing folks to configure this explicitly is a _really good thing_.

I'm kinda hoping to find a balance where it's possible to have a barebones config that invokes the `babel` plugin with no parameters when you're just getting started, and provides a really clear path towards a production-ready config.

Sometimes projects solve this with a bootstrap utility (`npm init`, for example). That could be an option, but IMO rollup's use-cases (mostly, source input and targets) are so varied that such a bootstrapping tool would be a very complex undertaking with too many moving parts to be useful.

I think I'm not in favor of a `mode`-type config item, but happy to be convinced otherwise. This is mostly on the grounds that I use rollup because it gives me a ton of power right out of the box, and doesn't make (too many) choices for me. This might inform our direction here, but I'm not sure if that's a common perspective.

>Maybe we could do similarly here - leave current logic (no config, bundled helpers) as default BUT give a short warning (with a link to a full explanation) about configuring this explicitly. WDYT?

I think this is more or less what I was attempting to say with "I'm tempted to say we should preserve some of the auto-detection functionality." If have the babel plugin run a preflight check to determine the appropriate `babelHelpers` key when not provided, and warn in that case, then I think that covers my concerns.

TLDR: :+1: on auto-detecting (no config, bundled helpers) as default with warning.

**Comment by skeggse** [comment]
Anything I can do to help move this along?

**Comment by Andarist** [comment]
@skeggse would you still be interested in helping to get this done?

**Comment by skeggse** [comment]
Yes!

**Comment by Andarist** [comment]
Could you fork from this branch, merge current master to it to resolve conflicts and resolve comments (the ones that are resolvable right now?

I could do a second round of code review after that, I suspect there wouldn't be much work left to do to by then so maybe we would be able to get this through a finish line without that much effort?

**Comment by skeggse** [comment]
Sure thing. I don't think I'll be able to slot it in until the end of the week, but I'll keep ya posted.

File: test/test.js (hunk 4 above)
Comment (skeggse) [reply test/test.js]: Yeah looks like this is because `buble/register` doesn't support options.

**Comment by skeggse** [comment]
Ok it's up at #325. I misread your comment earlier and rebased instead of merging current master into it. I can redo and go the merge route if you'd rather; let me know.

**Comment by Andarist** [comment]
Closing in favor of @skeggse's PR
